### DISCUSSION TITLE
Pocketbase and its scaling

### DISCUSSION THREAD

**Discussion by shaheer-haider** [pocketbase/pocketbase#2184] [Q&A]
I've some (maybe stupid) questions. Pocketbase is really good but in case it crashes or the server running it, crashes for any reason, there'll be nothing running in parallel. That's why I need a solution. I won't use any realtime connections, there'll be simple API calls.

1. Does it support caching?
2. Storing pd_data and pb_migrations in EBS while having pocketbase file running in an EC2, is it a good approach for scaling?
3. How much it can scale? I want it to be able to handle 100K users, there'll be many API calls, but If it doesn't support caching, I'll simple cache the responses in my Node server, but I still want idea from both perspectives.
4. How many connections it can handle on a gaint server and how much it can handle on multiple small servers? Just want a little bit of idea.
5. If having a single Sqlite file in EBS is not good approach then what is? Using Marmot?

Thank you.

**Comment by ganigeorgiev** [comment]
PocketBase out of the box can run only on a single server so it can scale only vertical.

How much - it depends. You can find some info about my earlier tests in https://github.com/pocketbase/pocketbase/discussions/1629#discussioncomment-4715328.
Official testbed and benchmarks will be available in the near future.

There is no built-in API cache. Note that for most use cases API cache shouldn't be necessary as it is already performant enough. In case you want to squeeze further the read/write query performance you can build PocketBase with `CGO_ENABLED=1`, which will enable the CGO SQLite driver. For some queries it could be 2-3x times faster. You can find benchmarks comparing the CGO (sqlite3) vs pure Go (sqlite) implementations in https://docs.google.com/spreadsheets/d/1YOP1D_ZhuR-ednQhTH6SSrCkcR5bOHzCDoQBZMXvyAw/edit (note that there is also a "Charts" sheet).

**Reply by shaheer-haider** [reply]
Thanks, @ganigeorgiev. Ac on these benchmarks, pocket base is really fast but I am still worried, what if the server crashes and the only server I had was that one, so?

**Reply by ganigeorgiev** [reply]
> but I am still worried, what if the server crashes and the only server I had was that one, so?

If the server crashes then it will restart.

If you are worried about hardware failure, then I don't think there is a "one size fit all" type of solution. The best approach in this case would be to have a backup somewhere else and to bring up a new server behind some load balancer. Depending on your use case you may also use something like [Litestream](https://litestream.io/) or [LiteFS](https://github.com/superfly/litefs) (both as far as I know support WAL mode) or as you mentioned earlier [Marmot](https://github.com/maxpert/marmot).

_Please also keep in mind that if your use case requires HA, "zero-downtime" deployments, etc. PocketBase may not be suitable for it (at least not out of the box and in its current development stage)._
